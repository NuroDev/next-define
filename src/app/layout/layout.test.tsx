import { describe, it, expect, expectTypeOf } from "vitest";
import renderer from "react-test-renderer";

import { defineLayout } from ".";

import type { ReactNode } from "react";

import type { FC } from "~/shared";
import type { NextLayoutProps } from "./layout.action";

describe("layout", () => {
  it("Create a basic layout wrapper component", () => {
    const { Component } = defineLayout({
      Component: ({ children, params }) => {
        expect(params).toBeDefined();
        expect(params).toBeInstanceOf(Object);
        expectTypeOf(children).toMatchTypeOf<ReactNode>();

        return <>{children}</>;
      },
    });

    const ComponentJson = renderer
      .create(<Component children={<></>} params={{}} />)
      .toJSON();
    expect(ComponentJson).toBeDefined();
    expect(ComponentJson).toMatchSnapshot();
    expectTypeOf(Component).toMatchTypeOf<FC<NextLayoutProps>>();
  });

  it("Create a basic layout wrapper component with typed parameters", () => {
    interface Params {
      foo: string;
    }

    const params = {
      foo: "bar",
    } satisfies Params;

    const { Component } = defineLayout<Params>({
      Component: ({ children, params }) => {
        expect(params).toBeDefined();
        expect(params).toBe(params);
        expectTypeOf(children).toMatchTypeOf<ReactNode>();

        return <>{children}</>;
      },
    });

    const ComponentJson = renderer
      .create(
        // @ts-expect-error Typed props means this always returns as a `FC` type & this function, like most, still don't support async components yet
        <Component children={<></>} params={params} />
      )
      .toJSON();
    expect(ComponentJson).toBeDefined();
    expect(ComponentJson).toMatchSnapshot();
    expectTypeOf(Component).toMatchTypeOf<FC<NextLayoutProps<Params>>>();
  });
});
