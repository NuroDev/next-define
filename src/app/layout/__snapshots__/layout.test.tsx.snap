// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layout > Create a basic layout wrapper component 1`] = `null`;

exports[`layout > Create a basic layout wrapper component with typed parameters 1`] = `null`;
